cmd/controller-manager/app/controllermanager.go

```diff
@@ -103,9 +103,238 @@ func Run(ctx context.Context, opts *options.Options) error {
 	return nil
 }
 
+// ControllerContext defines the context object for controller
+type ControllerContext struct {
+	Mgr                         controllerruntime.Manager
+	ObjectWatcher               objectwatcher.ObjectWatcher
+	Opts                        *options.Options
+	StopChan                    <-chan struct{}
+	DynamicClientSet            dynamic.Interface
+	OverrideManager             overridemanager.OverrideManager
+	ControlPlaneInformerManager informermanager.SingleClusterInformerManager
+}
+
+// InitFunc is used to launch a particular controller.  It may run additional "should I activate checks".
+// Any error returned will cause the controller process to `Fatal`
+// The bool indicates whether the controller was enabled.
+type InitFunc func(ctx ControllerContext) (enabled bool, err error)
+
+// NewControllerInitializers is a public map of named controller groups (you can start more than one in an init func)
+func NewControllerInitializers() map[string]InitFunc {
+	controllers := map[string]InitFunc{}
+	controllers["cluster"] = startClusterController
+	controllers["clusterStatus"] = startClusterStatusController
+	controllers["hpa"] = startHpaController
+	controllers["binding"] = startBindingController
+	controllers["execution"] = startExecutionController
+	controllers["workStatus"] = startWorkStatusController
+	controllers["namespace"] = startNamespaceController
+	controllers["serviceExport"] = startServiceExportController
+	controllers["endpointSlice"] = startEndpointSliceController
+	controllers["serviceImport"] = startServiceImportController
+	return controllers
```

Comment:
Can we just define this as a map? There is no other logic in this function either.

Reply:
I think so. We can define a map, and in other controller to add function to this map